src/ComponentProvider.ts

```diff
@@ -0,0 +1,199 @@
+            if (compareComponents(childComponent.json(), trait[String(key)])) {
+              return 0;
+            }
+            return 1;
+          }
+        }
+      }
+    }
+    return 2;
+  }
+
+  /**
+   * This function will scan a component by name from traits of another component.
+   *
+   * @param {string} name - the name of the component that needs to be extracted from traits.
+   * @param {any} child - the child that needs to be scanned.
+   * @param {any} parent - this is the parent. it should contain traits to be scanned.
+   * @param {string} path - path of the parent.
+   * @returns {void}
+   *
+   * @example
+   *
+   *     scanInTrait('payload', payload, message, path)
+   */
+  private scanInTrait = (name: string, child: any, parent: any, path: string): void => {
+    const traits = parent.extension('x-parser-original-traits');
+    for (let i = 0; i < traits.length; i++) {
+      for (const key in traits[parseInt(`${i}`, 10)]) {
+        if (key === name) {
+          this.scanSchema(`${path}.traits[${i}].${name}`, child);
+        }
+      }
+    }
+  }
+
+  /**
+   * This function will handle the cases which it is possible that the values of a
+   * field is coming from a trait.
+   *
+   * @param {string} name - the name of the component that needs to be extracted from traits.
+   * @param {string} parentPath - path of the parent.
+   * @param {any} child - the child that needs to be scanned.
+   * @param {any} parent - this is the parent. it should contain traits to be scanned.
+   * @returns {void}
+   *
+   * @example
+   *
+   *     handlePossibleInTraitsComponents('payload', path, payload, message)
+   */
+  private handlePossibleInTraitsComponents = (name: string, parentPath: string, child: any, parent: any) => {
+    switch (this.isInTraits(name, child, parent)) {
+    case 0:
+      this.scanInTrait(name, child, parent, parentPath);
+      break;
+    case 2:
+      this.scanSchema(`${parentPath}.${name}`, child);
+      break;
+    case 1:
+      // When we have mixed components that has fields in both traits and
+      // in component itself. we ignore this case for now.
+      break;
+    }
+  }
+  /**
+   * This function is responsible for scanning any message that is passed to it.
+   *
+   * @param {string} path - path of the message that needs to be scanned.
+   * @param {Message} message - the actual message object.
+   *
+   * @returns {void}
+   */
+  private scanMessage = (path: string, message: Message): void => {
+    const payload = message.payload();
+    const headers = message.headers();
+    if (payload) {
+      this.handlePossibleInTraitsComponents('payload', path, payload, message);
+    }
+    
+    if (headers) {
+      this.handlePossibleInTraitsComponents('headers', path, headers, message);
+    }
+  }
+  /**
+   * This function is responsible for scanning the document's components section.
+   *
+   * @returns {void}
+   */
+  private scanComponents = (): void => {
```

Comment:
Tip: you should avoid using the arrow function inside class, because then their definition is created for each instance - in opposite the normal function are saved to prototype of class. It can stay, but have it in mind for the future :)